RFC-0021-Distributed-Pipeline-Parallel-Technical.md

```diff
@@ -0,0 +1,620 @@
+* (**D7**) Unified programming model that already exists
+
+**Con**
+
+* (split pro/con) (**D2**) Not clear if `1f1b` or other schedules are implemented/implementable?
+* (**D8**) Not clear what the training loop abstraction looks like. The optimizer is installed via an `nn.Graph` API. Loss calculation is created in the `nn.Graph.build()` method.
+
+## Final Analysis
+
+### General Design Axes
+
+Looking through the various approaches above, we can pull out some general design axes we should consider:
+
+
+* (**DA1**) Single-coordinator vs. distributed coordination
+    * Sub-decisions: model execution, autograd, optimizers
+* (**DA2**) Python training loop vs. custom encapsulated training loop (e.g. as in DeepSpeed)
+* (**DA3**) `nn.Sequential` vs. more free-form partitioning
+* (**DA4**) Local vs. distributed loss
+* (**DA5**) Distributed autograd framework vs. manual autograd
+    * As it relates to pipeline schedules
+* (**DA6**) Local vs. distributed optimizer
+* (**DA7**) Synchronous pipeline dispatch vs. asynchronous
+* (**DA8**) Predication scheme (Approach 1 only)
+* (**DA9**) Instruction format (Approach 4 only)
+
+
+We can start analyzing the approaches by these design axes
+
+* Approach 1: SPMD with Predicated Training Loop and message passing
+* Approach 2: RPC with RemoteModule and torchgpipe-style single coordinator (@pritamdamania87 RFC)
+    * Single-coordinator
+        * CUDA command buffer analogy
+    * Continuation passing
+* Approach 3: MPMD with RemoteModule and message passing (fairscale experimental)
+* Approach 4: MPMD with a custom interpreter/instruction format and message passing (DeepSpeed)
+* Approach 5: RPC with remote modules and generalized Module-server architecture (SageMaker)
+* Approach 6: SPMD with Program capture/JIT compilation and message passing (OneFlow)
+
+|	|DA1	|DA2	|DA3	|DA4	|DA5	|DA6	|DA7	|DA8	|DA9	|Notes	|
+|---	|---	|---	|---	|---	|---	|---	|---	|---	|---	|---	|
+|Approach 1	|multi	|py	|FF	|local	|manual	|local	|async	|?	|X	|	|
+|Approach 2	|single	|py	|seq	|dist	|dist	|dist	|sync	|X	|X	|	|
+|Approach 3	|single	|py	|seq	|dist	|dist	|dist	|sync	|X	|X	|	|
+|Approach 4	|multi	|interp	|seq	|local?	|manual	|local	|async*	|X	|?	|	|
+|Approach 5	|single*	|py	|FF	|local?	|manual	|local?	|sync?	|X	|X	|Schedules?	|
+|Approach 6	|multi	|interp	|FF	|local?	|manual? (graph?)	|local?	|async?	|X	|X	|	|
+
+### Decision - Approach 3 with Modifications
+
+After deliberation, we want to build the API with the least complexity, at least initially. We will modify/build the API in FairScale experimental with a few modifications:
+
+
+* (**DA5**) Rather than using torchgpipe-style virtual dependencies in the distributed autograd graph, we want each stage to manually handle running `forward` and `backward` stages (the latter by explicitly calling `torch.autograd.backward()`). This will give easier and finer-grained control over the execution schedule of the pipeline
```

Comment:
Can an example of what the main training loop looks like with these modifications be added?

And also, how is training launched? Some example commands for each required rank / host would be useful.